[enh] use longest title and fix get_ordered_results()

allendema opened the pull request:
## What does this PR do?

<!-- MANDATORY -->

<!-- explain the changes in your PR, algorithms, design, architecture -->
Make sure that when multiple engines get the same result, the longest title will be used.
This was already done for 'description'.

## Why is this change important?

Some engines might return empty titles (mojeek). 
Not have the exact query at hand right now.


<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes -->
`make test`
`make run`
## Author's checklist



<!-- additional notes for reviewers -->
Without adding the initialize() there are a bunch of errors when using get_ordered_results().
Feel free to change anything.
## Related issues

<!--
Closes #234
-->
N/A

allendema committed:
[enh] use longest title and test get_ordered_results()

return42 commented on searx/results.py:
```diff
@@ -184,6 +183,7 @@ class ResultContainer:
 
     def __init__(self):
         super().__init__()
+        metrics_initialize()
```
Why is the init needed here .. as far I can say metrics should be initialized by the `initialize_metrics` call here in the `searx.search` package

https://github.com/searxng/searxng/blob/cf7627557ab29e31efbf4a0719d1083592e22242/searx/search/__init__.py#L36

allendema replied:
get_ordered_results() -> close() -> counter_add() -> counter_storage not defined (or similiar) would somehow fail.

 So I just made sure that counter_storage was set by initializing it it explicitly. I'm sure there is another way to solve it.

return42 replied:
> get_ordered_results() -> close() -> counter_add() -> counter_storage not defined (or similiar) would somehow fail.

Sure? .. I can't reproduce it .. I commented out `metrics_initialize` and made some tests `make test run ..` without an issue.

allendema replied:
Something is flaky.
Here it's reproducible and this is one example traceback.

```ERROR: test_result_merge_by_title (tests.unit.test_results.ResultContainerTestCase.test_result_merge_by_title)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspaces/searxng/tests/unit/test_results.py", line 72, in test_result_merge_by_title
    self.container.extend('stract', [fake_result(engine='stract', title='short title')])
  File "/workspaces/searxng/searx/results.py", line 243, in extend
    histogram_observe(standard_result_count, 'engine', engine_name, 'result', 'count')
  File "/workspaces/searxng/searx/metrics/__init__.py", line 49, in histogram_observe
    histogram_storage.get(*args).observe(duration)
    ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get'
```

return42 replied:
This may be due to the unit tests ... they are not always executed in the same order ... for me metrics seems to be already initialized when the test_results.py is executed ... for you the order is different and then the initialization does not seem to have taken place yet ... our tests do not build and unbuild their environments cleanly ... I will have to take a closer look at the problem ...

return42 replied:
@allendema I removed the `metrics_initialize()`  from your commit and I placed a commit on top of your branch to fix the issues you have seen in the unit tests, further I rebased you branch on current master.

Can you please give it another try .. thanks :+1:

return42 replied:
Argh .. unit tests fails in the CI .. seems I have to do more work / comming soon ..

return42 committed:
[fix] unit tests: call searx.search.initialize in test's setUp

Depending on the order the unit tests are executed, the searx.search module is
initalized or not, issue reported in [1]::

    Traceback (most recent call last):
      File "searxng/tests/unit/test_results.py", line 72, in test_result_merge_by_title
        self.container.extend('stract', [fake_result(engine='stract', title='short title')])
      File "searxng/searx/results.py", line 243, in extend
        histogram_observe(standard_result_count, 'engine', engine_name, 'result', 'count')
      File "searxng/searx/metrics/__init__.py", line 49, in histogram_observe
        histogram_storage.get(*args).observe(duration)
        ^^^^^^^^^^^^^^^^^^^^^
      AttributeError: 'NoneType' object has no attribute 'get'

To ensure that the searx.search module is initialized, the

- searx.engines.load_engines is replace by
- searx.search.initialize

[1] https://github.com/searxng/searxng/pull/3932#discussion_r1822406569

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 reviewed (approved):
Longer title is shown in the result & CI has now been passed ;-) .. Thanks for this PR :+1:
